File: post_service/python/controllers/postController.py
from flask import Blueprint, request, jsonify # type: ignore
from bson.objectid import ObjectId # type: ignore
from pymongo import MongoClient # type: ignore
from datetime import datetime

client = MongoClient('mongodb://localhost:27017/')
db = client['post_database']
posts = db['posts']

post_blueprint = Blueprint('posts', __name__)

@post_blueprint.route('/posts', methods=['POST'])
def create_post():
    data = request.get_json()
    user_id = data.get('user_id')
    content = data.get('content')
    media = data.get('media', [])

    post = {
        "post_id": ObjectId(),
        "user_id": ObjectId(user_id),
        "content": content,
        "media": media,
        "likes": 0,
        "comments": [],
        "created_at": datetime.utcnow()
    }

    result = posts.insert_one(post)
    new_post = posts.find_one({"_id": result.inserted_id})

    return jsonify(new_post), 201

@post_blueprint.route('/posts', methods=['GET'])
def get_all_posts():
    all_posts = posts.find()
    post_list = [post for post in all_posts]
    return jsonify(post_list), 200

@post_blueprint.route('/posts/<post_id>', methods=['GET'])
def get_post_by_id(post_id):
    post = posts.find_one({"post_id": ObjectId(post_id)})

    if not post:
        return jsonify({"message": "Post not found"}), 404

    return jsonify(post), 200

@post_blueprint.route('/posts/<post_id>/comments', methods=['POST'])
def add_comment(post_id):
    data = request.get_json()
    user_id = data.get('user_id')
    content = data.get('content')

    comment = {
        "comment_id": ObjectId(),
        "user_id": ObjectId(user_id),
        "content": content,
        "created_at": datetime.utcnow()
    }

    result = posts.update_one(
        {"post_id": ObjectId(post_id)},
        {"$push": {"comments": comment}}
    )

    if result.matched_count == 0:
        return jsonify({"message": "Post not found"}), 404

    post = posts.find_one({"post_id": ObjectId(post_id)})
    return jsonify(post), 200

File: services/post_service/controllers/postController.js
const Post = require('../models/postModel');

// Create a new post
const createPost = async (req, res) => {
    const { user_id, content, media } = req.body;

    try {
        const post = new Post({
            user_id,
            content,
            media,
            created_at: new Date()
        });

        await post.save();
        res.status(201).json(post);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// Get all posts
const getAllPosts = async (req, res) => {
    try {
        const posts = await Post.find();
        res.status(200).json(posts);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// Get a post by ID
const getPostById = async (req, res) => {
    const { id } = req.params;

    try {
        const post = await Post.findById(id);

        if (!post) {
            return res.status(404).json({ message: 'Post not found' });
        }

        res.status(200).json(post);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// Add a comment to a post
const addComment = async (req, res) => {
    const { id } = req.params;
    const { user_id, content } = req.body;

    try {
        const post = await Post.findById(id);

        if (!post) {
            return res.status(404).json({ message: 'Post not found' });
        }

        post.comments.push({ user_id, content, created_at: new Date() });
        await post.save();

        res.status(200).json(post);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

module.exports = {
    createPost,
    getAllPosts,
    getPostById,
    addComment
};
